src/sources/util/decoding/mod.rs

```diff
@@ -0,0 +1,196 @@
+mod config;
+mod framing;
+mod parsers;
+
+use crate::{event::Event, internal_events::DecoderParseFailed, sources::util::TcpError};
+use bytes::{Bytes, BytesMut};
+pub use config::{DecodingConfig, FramingConfig};
+pub use framing::OctetCountingDecoder;
+pub use parsers::BytesParser;
+#[cfg(feature = "sources-syslog")]
+pub use parsers::SyslogParser;
+use tokio_util::codec::LinesCodecError;
+
+pub trait Parser {
+    fn parse(&self, bytes: Bytes) -> crate::Result<Event>;
+}
+
+pub trait DecoderError: std::error::Error + TcpError + Send + Sync {}
+
+#[derive(Debug)]
+pub struct Error {
+    error: std::sync::Arc<dyn DecoderError>,
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(formatter, "{}", self.error)
+    }
+}
+
+impl TcpError for Error {
+    fn is_fatal(&self) -> bool {
+        self.error.is_fatal()
+    }
+}
+
+impl DecoderError for std::io::Error {}
+
+impl DecoderError for LinesCodecError {}
+
+impl From<std::io::Error> for Error {
+    fn from(error: std::io::Error) -> Self {
+        Self {
+            error: std::sync::Arc::new(error),
+        }
+    }
+}
+
+impl From<LinesCodecError> for Error {
+    fn from(error: LinesCodecError) -> Self {
+        Self {
+            error: std::sync::Arc::new(error),
+        }
+    }
+}
+
+pub struct Decoder {
+    framer: Box<dyn tokio_util::codec::Decoder<Item = Bytes, Error = Error> + Send + Sync>,
```

Comment:
We're repeating this type a lot and it's pretty complex. Should we define our own trait like we did for `DecoderError`?

Reply:
This has been updated accordingly:

https://github.com/timberio/vector/blob/e1065d4e30975db3ff31cd93e372a1d19d1d8070/src/sources/util/decoding/mod.rs#L90-L98